src/scripts/shards/Shards.ts

```diff
@@ -99,6 +134,14 @@ class Shards implements Feature {
         if (json != null) {
             this.shardWallet = getArrayOfObservables(json['shardWallet']);
             this.shardUpgrades = getArrayOfObservables(json['shardUpgrades']);
+            // Refund invalid upgrades
+            for (let t = 0; t < Shards.nTypes; t += 1) {
+                for (let effect = 0; effect < Shards.nEffects; effect += 1) {
+                    if (this.isInvalidUpgrade(t, effect)) {
+                        this.refundShardUpgrade(t, effect);
+                    }
+                }
+            }
```

Comment:
Any updating of save data should be done in the [`Update.ts` file](https://github.com/pokeclicker/pokeclicker/blob/develop/src/scripts/Update.ts).